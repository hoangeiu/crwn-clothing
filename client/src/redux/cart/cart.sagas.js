import { all, call, takeLatest, put, select } from "redux-saga/effects";

import UserActionTypes from "../user/user.types";
import CartActionTypes from "./cart.types";
import { clearCart, setCartFromFireBase } from "./cart.actions";
import { getUserCartRef } from "../../firebase/firebase.utils";
import { selectCartItems } from "./cart.selectors";
import { selectCurrentUser } from "../user/user.selector";

export function* clearCartOnSignOut() {
  yield put(clearCart());
}

export function* checkCartFromFirebase({ payload: user }) {
  const cartRef = yield call(getUserCartRef, user.id);

  const cartSnapshot = yield cartRef.get();

  yield put(setCartFromFireBase(cartSnapshot.data().cartItems));
}

export function* updateCartToFirebase() {
  const currentUser = yield select(selectCurrentUser);
  if (currentUser) {
    try {
      const cartRef = yield call(getUserCartRef, currentUser.id);

      const cartItems = yield select(selectCartItems);

      yield cartRef.update({ cartItems });
    } catch (error) {
      console.log(error);
    }
  }
}

export function* onSignOutSuccess() {
  yield takeLatest(UserActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOut);
}

export function* onUserSignIn() {
  yield takeLatest(UserActionTypes.SIGN_IN_SUCCESS, checkCartFromFirebase);
}

export function* onUserUpdateCart() {
  yield takeLatest(
    [
      CartActionTypes.ADD_ITEM,
      CartActionTypes.REMOVE_ITEM,
      CartActionTypes.CLEAR_ITEM_FROM_CART,
    ],
    updateCartToFirebase
  );
}

export function* cartSagas() {
  yield all([
    call(onSignOutSuccess),
    call(onUserSignIn),
    call(onUserUpdateCart),
  ]);
}
