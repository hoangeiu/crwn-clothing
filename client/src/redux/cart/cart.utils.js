export const addItemToCart = (cartItems, itemToAdd) => {
  const existingCartItem = cartItems.find((item) => item.id === itemToAdd.id);

  if (existingCartItem) {
    return cartItems.map((item) =>
      item.id === itemToAdd.id ? { ...item, quantity: item.quantity + 1 } : item
    );
  }

  return [...cartItems, { ...itemToAdd, quantity: 1 }];
};

export const removeItemFromCart = (cartItems, itemToRemove) => {
  const existingCartItem = cartItems.find(
    (item) => item.id === itemToRemove.id
  );

  if (existingCartItem.quantity === 1)
    return cartItems.filter((cartItem) => cartItem.id !== itemToRemove.id);

  return cartItems.map((cartItem) =>
    cartItem.id === itemToRemove.id
      ? { ...cartItem, quantity: cartItem.quantity - 1 }
      : cartItem
  );
};
