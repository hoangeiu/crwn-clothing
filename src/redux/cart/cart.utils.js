export const addItemToCart = (cartItems, itemToAdd) => {
  const existingCartItem = cartItems.find((item) => item.id === itemToAdd.id);

  if (existingCartItem) {
    return cartItems.map((item) =>
      item.id === itemToAdd.id ? { ...item, quantity: item.quantity + 1 } : item
    );
  }

  return [...cartItems, { ...itemToAdd, quantity: 1 }];
};
