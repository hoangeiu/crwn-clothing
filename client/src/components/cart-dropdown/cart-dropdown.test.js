import React from "react";
import { shallow } from "enzyme";

import { CartDropdown } from "./cart-dropdown.component";
import CartItem from "../cart-item/cart-item.component";

import { toggleCartHidden } from "../../redux/cart/cart.actions";

describe("CartDropdown component", () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];
  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch,
    };

    wrapper = shallow(<CartDropdown {...mockProps} />);
  });

  it("should CardDropdown renders without error", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render an equal number of CardItem components as the cardItems prop", () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });

  it("should render empty message if cardItems is empty", () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch,
    };

    wrapper = shallow(<CartDropdown {...mockProps} />);

    expect(wrapper.exists(".empty-message")).toBe(true);
  });

  it("should call history.push when button is clicked", () => {
    wrapper.find("CustomButton").simulate("click");
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });
});
