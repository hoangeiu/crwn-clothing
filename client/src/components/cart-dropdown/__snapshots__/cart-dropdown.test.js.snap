// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component should CardDropdown renders without error 1`] = `
<div
  className="cart-dropdown"
>
  <div
    className="cart-items"
  >
    <Memo(CartItem)
      item={
        Object {
          "id": 1,
        }
      }
      key="1"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 2,
        }
      }
      key="2"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 3,
        }
      }
      key="3"
    />
  </div>
  <CustomButton
    onClick={[Function]}
  >
    GO TO CHECKOUT
  </CustomButton>
</div>
`;
